src/main/java/com/zufar/onlinestore/cart/api/CartApiImpl.java

```diff
@@ -0,0 +1,60 @@
+package com.zufar.onlinestore.cart.api;
+
+import com.zufar.onlinestore.cart.converter.ShoppingSessionDtoConverter;
+import com.zufar.onlinestore.cart.dto.AddNewItemToShoppingSessionRequest;
+import com.zufar.onlinestore.cart.dto.GetShoppingSessionRequest;
+import com.zufar.onlinestore.cart.dto.RemoveItemFromShoppingSessionRequest;
+import com.zufar.onlinestore.cart.dto.ShoppingSessionDto;
+import com.zufar.onlinestore.cart.dto.UpdateProductsQuantityInShoppingSessionItemRequest;
+import com.zufar.onlinestore.cart.entity.ShoppingSession;
+import com.zufar.onlinestore.cart.entity.ShoppingSessionItem;
+import com.zufar.onlinestore.cart.repository.ShoppingSessionRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Objects;
+
+@Service
+@RequiredArgsConstructor
+public class CartApiImpl implements CartApi {
+    
+    private final ShoppingSessionRepository shoppingSessionRepository;
+    private final ShoppingSessionDtoConverter shoppingSessionDtoConverter;
+
+    @Override
+    public ShoppingSessionDto getShoppingSession(final GetShoppingSessionRequest getShoppingSessionRequest) {
+        return null;
+    }
+
+    @Override
+    public ShoppingSessionDto addNewItemToShoppingSession(final AddNewItemToShoppingSessionRequest addNewItemToShoppingSessionRequest) {
+        return null;
+    }
+
+    @Override
+    public ShoppingSessionDto removeItemFromShoppingSession(final RemoveItemFromShoppingSessionRequest removeItemFromShoppingSessionRequest) {
+        return null;
+    }
+
+    @Override
+    @Transactional(propagation = Propagation.REQUIRED, isolation = Isolation.READ_COMMITTED)
+    public ShoppingSessionDto updateProductAmountInShoppingSessionItem(final UpdateProductsQuantityInShoppingSessionItemRequest request) {
+        ShoppingSession shoppingSession = shoppingSessionRepository.findById(request.shoppingSessionId())
+                .orElseThrow();
+
+        ShoppingSessionItem shoppingSessionItem = shoppingSession.getItems().stream()
+                .filter(item -> Objects.equals(item.getId(), request.shoppingSessionItemId()))
+                .findFirst()
+                .orElseThrow();
+
+        Integer newProductsQuantity = shoppingSessionItem.getProductsQuantity() + request.productsQuantityChange();
+        shoppingSessionItem.setProductsQuantity(newProductsQuantity);
```

Comment:
It is not a good idea to overwrite value at all. Better to change value as a patch using custom query.
In your case hibernate will generate query like this 

update shopping_session_item set products_quantity = :products_quantity where id = :id

If two request come from user at the same moment (it is very likely in this case) then we will have lost update problem

I will try to explain:
Let's imagine that in the database we have actual value of "products_quantity" = 10

Let's imagine first transaction gets value of shoppingSession.getItems() (and one of this items has "products_quantity" = 10)
and then wants to change +5 and calculating
newProductsQuantity = shoppingSessionItem.getProductsQuantity() + request.productsQuantityChange();
newProductsQuantity  = 15  = 10 + 5 

At the same time second transaction gets the same value of shoppingSession.getItems() (also"products_quantity" = 10) (because first transaction is not yet committed) and second transaction sees only committed values (READ_COMMITED)
And then wants to change +3 and calculating 
newProductsQuantity = shoppingSessionItem.getProductsQuantity() + request.productsQuantityChange();
newProductsQuantity  = 13 = 10 + 3

then first transaction finally updates the row in database
update shopping_session_item set products_quantity = 15 where id = :id

then second transaction also updates the row in database
update shopping_session_item set products_quantity = 13 where id = :id

At the end of the transaction in the database we have actual value = 15 or 13, it depends on which query will be faster, but 15 and 13 is not correct values because we should have all changes and should have value 18 = 10 + 5 + 3


First solution is to write custom query 
@Query("update shopping_session_item
  set products_quantity = products_quantity +  :products_quantity_change where id = :id")
    @Modifying
    void  updateProductAmountInShoppingSessionItem(@Param("id") Long Id,
                 @Param("products_quantity_change") Long productsQuantityChange);
                 
Second solution is just to change isolation level - @Transactional(isolation = Isolation.REPEATABLE_READ)  but in this case second transaction will be aborted due to concurrent update and you should retry second one. 
If you have 10 transactions trying to update the same row at the same time then 9 of 10 transactions will be aborted. 
I mean only 1 of N concurrent transactions will be commited.

I think first solution is better and more understandable.

Reply:
Fixed